### DISCUSSION TITLE
Nextcloud Compatibility

### DISCUSSION THREAD

**Discussion by Opisek** [Opisek/luna#3] [General]
Opening this thread as a place to aggregate information and issues about Nextcloud CalDav server compatibility throughout the development process.

@tiritibambix

**Comment by Tiritibambix** [comment]
The base URL given by nextcloud is ```https://mydomain.click/remote.php/dav```
The correct way to add it to Luna is ```https://mydomain.click/remote.php/dav/calendars/username/```

I have 4 calendars in my NC calendar. It won't fetch one of them for this reason:
```
2024-12-09T15:10:03.244469748Z ERRO[2024-12-09T16:10:03+01:00] could not get events: could not get events from calendar Jours fériés: could not convert event /remote.php/dav/calendars/admin/jours-fris/5692E1DC-A283-458D-B0F2-62F91A819D3F.ics: event has no end time or duration  module=api
```

It won't create a new event for this reason:
```
2024-12-09T15:12:33.171782043Z ERRO[2024-12-09T16:12:33+01:00] could not add event: could not add event: 415 Unsupported Media Type: <error xmlns="DAV:"><exception xmlns="http://sabredav.org/ns">Internal Server Error</exception><message xmlns="http://sabredav.org/ns">

2024-12-09T15:12:33.171853523Z &#x9;&#x9;The server was unable to complete your request.&#x9;&#x9;If this happens again, please send the technical details below to the server administrator.&#x9;&#x9;More details can be found in the server log.&#x9;&#x9;&#x9;</message><technical-details xmlns="http://sabredav.org/ns">
2024-12-09T15:12:33.171871023Z &#x9;&#x9;<remote-address xmlns="http://sabredav.org/ns">myipaddress</remote-address>
2024-12-09T15:12:33.171882980Z &#x9;&#x9;<request-id xmlns="http://sabredav.org/ns">Pca1uPkrRPYBpXAGphy2</request-id>
2024-12-09T15:12:33.171894462Z 
2024-12-09T15:12:33.171904084Z &#x9;&#x9;</technical-details></error>  module=api
``` 

_Originally posted by @tiritibambix in https://github.com/Opisek/luna/discussions/2#discussioncomment-11509799_

**Reply by Opisek** [reply]
Could you manually check what the event that fails is displayed as in Nextcloud? Perhaps they way they save all-day events there is by not storing an end date?

**Reply by Tiritibambix** [reply]
About this first error ```event has no end time or duration```, I believe this is due to this calendar being created by a premade vcs I downloded and imported to NC that displays all public holidays in the year, every year. That's why it has no end time I guess.

**Reply by Opisek** [reply]
I've encountered the fetch issue it in some other public holidays source myself. This can be seen in some events that last exactly one whole day. This does not conform to the specification, but I have added support for it to Luna.

(There is still a bug regarding this. If that event just so happens to be on a daylight savings time switch, the frontend breaks a little for that particular event. Vikunja no.122)

**Reply by Tiritibambix** [reply]
I just tried it and I don't have an error fetching this particular calendar. Good job :)

**Comment by Tiritibambix** [comment]
Since the last update, I've been having trouble fetching some calendars. For some reason, out of my four calendars, only the one dedicated to public holidays is being fetched.

Backend logs: [https://clip.steph.click/-N2ymYHppbV](https://clip.steph.click/-N2ymYHppbV)

At one point, the backend turns `unhealthy`, and I receive the error: `Failed to fetch events: Database error`.

**Reply by Opisek** [reply]
Can you manually increase `3*time.Second` to something higher like `30*time.Second` in [here](https://github.com/Opisek/luna/blob/main/backend/src/api/server.go) and rebuild?

**Reply by Opisek** [reply]
Repull before doing that. There's a fix to the `Database error`. It was caused by database connections not being closed cleanily on request timeout, despite everyone everywhere telling me that it should be handled automatically.

**Reply by Tiritibambix** [reply]
> Can you manually increase `3*time.Second` to something higher like `30*time.Second` in [here](https://github.com/Opisek/luna/blob/main/backend/src/api/server.go) and rebuild?

That worked. It fetches all calendars now, thank you.
I had an issue while creating an event though:
`ERRO[2025-04-05T19:57:52+02:00] Could not add event: Remote server returned an error  module=api`

I suspected it was due to the Nextcloud server I'm temporarily using, of which I am not an administrator. However, I tried with Bloben, which I still use for now, and it worked.

**Reply by Opisek** [reply]
You left some more details about event creation further up. This is still something that needs to be worked on, but it might take me a while due to not using Nextcloud myself.

I am very surprised that you needed to extend the timeout duration! I guess I will raise it to 5 seconds by default and add an option to adjust it via environmental variables. Since you're not the administrator of the Nextcloud server, I guess it is far away from you, which causes delay and a subsequent timeout.

**Reply by Opisek** [reply]
Pushed, you can set it in the backend as an environmental variable: ``REQUEST_TIMEOUT_DEFAULT=5``

**Reply by Tiritibambix** [reply]
All good. Thanks a lot

**Reply by Opisek** [reply]
FYI: I raised the default to 15s because Google Calendar takes some time when you try to add a calendar to it.
